### DISCUSSION TITLE
Zoom and panning in editor

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#13189] [Ideas]
**Describe the new feature:** 
I propose the ability to zoom and pan the view in the editor compose screen. This involves being able to zoom in to a smaller area in the editor so you can work on small details and panning so you can drag the playfield to look at another area while zoomed in. 
Panning should also be able to go offscreen so you can reach offscreen elements like far-off bezier anchors.
The screen size of slider anchors should stay the same when zooming in.

**Why this is usefull:**
Making small details in sliders has always been quite difficult without zoom. The size of the slider anchors in lazer makes this even harder, because they are large and hard to make out when overlapped. A zoom feature would make this much easier.

Making large sliders would also become easier if you can pan offscreen. Often when making bezier sliders the anchors need to be offscreen, because the footprint of the interpolated bezier path is always smaller than the footprint of the anchors. 

If there ever comes a rectangle outside the playfield that helps visualize the boundary between onscreen and offscreen, then panning and zooming out can be used be modders to check for offscreen objects.

**Proposal designs of the feature:** 

- A hotkey that when pressed lets you click and drag to pan and scroll with the mousewheel to zoom. Doubleclick to reset.
- Another option in the Toolbox that lets you control the camera. With the tool selected you can control the zoom and panning with the mouse just like described above.

**Comment by peppy** [comment]
Arguable how helpful this would be. For instance, "offscreen bezier nodes" should not be a thing given the new curve types. Zooming in maybe useful, but I'd honestly need to see an example where this is actually a thing that would improve the UX.

Adding something like this is easy, but I'm hesitant to complicate the UX for something which may only be used by one in 10,000 beatmaps.

**Comment by Walavouchey** [comment]
While I can't argue for slider composition, I can say this would at least be useful for a future design editor, where you can inspect and move storyboard objects in from offscreen. If something like this was implemented in the editor now, it would be immediately useful for modding storyboards for the same reason (once they can be viewed in the editor alongside hitobjects).

**Comment by OliBomby** [comment]
> Arguable how helpful this would be. For instance, "offscreen bezier nodes" should not be a thing given the new curve types. 

Even with the new curve types bezier segments will be used a lot, because they make the best smooth continuous curves. I can vouch with my personal experience of creating sliders it occurs very often that a few bezier anchors have to be moved offscreen and then I had to move the entire slider so the anchor comes onscreen, move the anchor, and then move the slider back to its original position.

> Zooming in maybe useful, but I'd honestly need to see an example where this is actually a thing that would improve the UX.

Mappers often put smaller details in sliders that involve 2 or 3 anchors very close to eachother. An example of this is called a "tumor" and it gets used in a lot of maps. Also high CS mapping has even smaller objects. Try mapping in CS 10 without zooming in.

Adding onto the point by @Walavouchey, I think zooming and panning is a very basic feature any editor should have. Zooming and panning might be essential for other future gamemodes using the same editor.


I'll further my point with some concrete examples:

Example of a slider with a tumor. The anchors are overlapping and its hard to select the anchor on the bottom of the stack. It's also impossible to see the exact centres of the anchors, so you wouldn't know if the anchors are laid out in a triangle which follows the broader curve of the slider.
![image](https://user-images.githubusercontent.com/17460441/113422254-a6718100-93cc-11eb-9bd0-2638b44be915.png)

Some other small details with 2 anchors close to eachother which is not tumors.
![image](https://user-images.githubusercontent.com/17460441/113422617-4fb87700-93cd-11eb-8433-f8d80694cb5a.png)

CS 10 mapping. Trying to create a common blanket has me squinting 5 cm away from my monitor.  The slider anchors also became really tiny and hard to select. 
![image](https://user-images.githubusercontent.com/17460441/113422610-4b8c5980-93cd-11eb-83f3-1a35c2472c58.png)

A slider in the shape of a pretzel with one continuous curve. The anchors have to be far offscreen. Its impossible to edit without scaling down the slider first, at which point you cant properly see the path of the slider.
![image](https://user-images.githubusercontent.com/17460441/113422691-6f4f9f80-93cd-11eb-8f2e-5749bdd85deb.png)

Slider with a little loop on the outside of the playfield. The bezier anchors overlap with the timeline and become impossible to select without moving the entire slider first. The last anchor is also impossible to select because it overlaps with the flip vertical button.
![image](https://user-images.githubusercontent.com/17460441/113422801-9c03b700-93cd-11eb-8ebe-db77594bd21a.png)

**Comment by peppy** [comment]
I'm already aware of such blasphemous mapping, but all of those sliders can be improved using perfect curves.

As for "off screen" objects: that should not be allowed. It's bad UX. we should not be adding a zoom out to let the user find invalid hit objects. they should be removed at the point of editing the beatmap.

No further examples are required, this can be left as a proposal for the distant future.

**Comment by alexpado** [comment]
I don't know if it's the right spot but I'd like to get back on the zoom feature when mapping in mania. 

This is something I use a lot in the stable version but that I found missing in lazer, which makes mapping way more comfortable (having a global view of a section / zooming in on a burst section, but also depending on the snap divisor).

We can change the zoom on the waveform but not the mapping area for now
